RFC: new tlb cap restrictions

nwf opened the pull request:
A first stab at the "revocation 3.0" design which requires that `CLC` to trap on loads (optionally, in a data-dependent way) if a per-TLB-entry 1-bit generation counter mismatches a global one.  This is probably not idiomatic `sail`, and all comments are more than welcome.

nwf committed:
Remove CCLC

Entirely too experimental

nwf committed:
Proposed extended TLB capability restriction taxonomy

nwf committed:
Introduce trap-on-cap-load behaviour

jrtc27 commented on cheri/cheri_insts.sail:
```diff
@@ -1259,11 +1259,13 @@ function clause execute (CSC(cs, cb, rt, offset)) =
+        Clear => false,
+        Trap  => raise_c2_exception(CapEx_TLBNoStoreCap, cs)
+      };
+      MEMw_tagged(pAddr, cap_size, cs_val.tag, capToMemBits(cs_val))
```
Should be `mtag` not `cs_val.tag`?

rmn30 commented on cheri/cheri_prelude_common.sail:
```diff
@@ -149,7 +149,7 @@ function readCapRegDDC(n) =
 
 /*!
-THIS`(cd, cap_val)` writes capability, `cap_val` capability register `cd`. Writes to register zero are ignored.
+THIS`(cd, cap\_val)` writes capability, `cap\_val` capability register `cd`. Writes to register zero are ignored.
```
This used to work, what has changed?

rmn30 replied:
This works for me so I'm curious what the problem is. A better fix might be to rename `cap_val` to `capVal` -- the code is already very inconsistent about camel case vs. snake case :-(

rmn30 commented on mips/mips_prelude.sail:
```diff
@@ -342,7 +342,7 @@ function rGPR idx = {
 
 /*!
-THIS(rd, v) writes the 64-bit value, _v_, to the general purpose register _rd_. Writes to register zero are ignored.
+THIS(rd, v) writes the 64-bit value, v, to the general purpose register rd. Writes to register zero are ignored.
```
These underscores should cause sail to emit `\emph{...}` for affected region. How are you generating latex from sail?

rmn30 commented on mips/mips_prelude.sail:
```diff
@@ -93,19 +94,21 @@ let MAX_U64 = MAX(64) /*unsigned(0xffffffffffffffff)*/
-  c1       : 35..  33,
-  d1       : 32,
-  v1       : 31,
+bitfield TLBEntry : bits(119) = {
```
Sail used to require registers to have bits type but this is no longer the case so this could probably be a struct and we wouldn't have to re-number all the fields. Possible refactoring for future.

rmn30 commented:
Looks like TLBEntryHi.CLG is never written. Probably need to modify MTC0 code, function TLBWriteEntry and TLBR -- search for all uses of TLBEntryHi

nwf commented:
`TLBEntryHi.CLG` is now modified as part of `MTC0` but is not modified when the TLB itself updates `TLBEntryHi`.  I'm not a fan of this whole "current state in the `TLBEntryHi`" thing, but I guess it is what it is.

rmn30 reviewed (approved):
Looksk good now.

bauereiss reviewed (approved):
Looks good to me from a verification perspective.

jonwoodruff reviewed (approved):
I didn't dig into the details, but the high-level sounds like the most reasonable approach for MIPS.

nwf commented:
From slack, we think we might want two bits in the EntryLo*: a per-entry CLG-mismatch-traps enable that we can set on the user page entries and not kernel page entries, so that the kernel doesn't accidentally its own foot when clocking the global CLG forward.  We might, possibly, find that this can be conflated with the existing Global bit, so that Global entries don't fire CLG traps.

Robert also asks about the hypervisor case, and that probably deserves some thought.

I know that @brettferdosi is champing at the bit for spec writing, so hopefully he and I can put our heads together this upcoming week and get some prose in the ISA doc for this as well as more fully flesh out the super-/hyper-visor story.

nwf committed:
CLoadTags: interaction with TLB restrictions

Trap if either the load inhibit is set (credit to James for the
suggestion) or if the load generation counters don't match.

nwf committed:
MIPS: refactor TLB segmentation decode

nwf commented:
Here's a version that includes @jonwoodruff's suggestion that we use the segment being indexed (User/Kernel/Supervisor) to decide between different global bits in EntryHi.  I'm pretty sure we never intend for a page to be multiply mapped across segments; that seems like it'd be a strange thing to do.

I've also added some traps for CLoadTags using the new TLB capability restriction taxonomy.

rmn30 commented on cheri/cheri_insts.sail:
```diff
@@ -1140,13 +1140,24 @@ function clause execute (CLoadTags(rd, cb)) =
+        match macr {
+
+            /* XXX Probably not the right exception cause */
+          Clear => raise_c2_exception(CapEx_TagViolation, cb),
```
Perhaps TLB Load cap would make more sense since it is ultimately still a TLB issue? Either that or another (rather specific) exception code.

rmn30 commented on mips/mips_prelude.sail:
```diff
@@ -66,6 +67,7 @@ bitfield TLBEntryLoReg : bits(64) = {
 
 bitfield TLBEntryHiReg : bits(64) = {
   R : 63.. 62,
+  CLG : 61.. 59, /* CHERI: Capability Load Generation: Kernel, Supervisor, User */
```
Could separate this field into CLGK, CLGS and CLGU?

nwf replied:
That's probably a good idea.

rmn30 commented on mips/mips_tlb.sail:
```diff
@@ -58,12 +58,36 @@ function tlbSearch(VAddr) =
     None()
   }
 
+val TLBSegmentOf : bits(64) -> (AccessLevel, option(bits(64))) effect pure
```
Is this function unmodified from the original? Fine to factor it out but maybe MIPSSegmentOf would be a more appropriate name since, of course if we are being picky the TLB is not used at all for the direct-mapped regions!

nwf replied:
It's not modified from the original; I'd done the refactoring as part of preparing for a change that I ended up dropping.  Will rename.

rmn30 reviewed (approved):
Fine just a few minor comments.

rmn30 committed:
Merge branch 'master' into new-tlb-cap-restrictions

rmn30 commented:
I had to resolve a large conflict because of recent commit that changed all decode lines. Thankfully changes on this branch only removed a line so think I've done it correctly.
